projects/elements/src/lib/lazy-elements/lazy-elements-loader.service.ts

```diff
@@ -18,14 +23,19 @@ export interface ElementConfig {
   providedIn: 'root'
 })
 export class LazyElementsLoaderService {
-  registry: Map<string, Promise<void>> = new Map<string, Promise<void>>();
   configs: ElementConfig[] = [];
 
   constructor(
+    @Optional()
```

Comment:
Personally, I am not really convinced of using `InjectionToken` for this case, 
if we move to `FactoryProvider` or `ClassProvider`. 🤔 

```typescript
module.forRoot({ 
 ...config,
lazyElementsStateProvider: {
  provide: LazyElementsStateService,
  userClass: DefaultLazyElementsStateService
}
})
```

or

```typescript
module.forRoot({ 
 ...config,
lazyElementsStateProvider: {
  provide: LazyElementsStateService,
  useFactory: myFactory,
  deps: [...myDependcies]
}
})
```

After doing that, we could get rid of the `optional` decorator & the empty check since by default we will use default state service.

@ye3i what do you think?

Reply:
I agree that it would be worth to provide configuration and providers in one place. 
But this approach will require that you import the library with the forRoot() method at least once.

```suggestion
    imports: [
      LazyElementsModule.forFeature()
  ]
```
If that's ok, then I can change my approach.

Reply:
On the other hand, we can just use the default provider replacement.

LazyElements.module
```
imports: [
{ provide: LazyElementsStateService, useClass: DefaultStateService }
]
```



App.module
```
imports: [
LazyElementsModule.forRoot(),
{ provide: LazyElementsStateService, useClass: CustomStateService }
]
```

Feature.module
```
imports: [
LazyElementsModule.forFeature(),
{ provide: LazyElementsStateService, useClass: CustomStateService }
]
```

Child.module
```
imports: [
LazyElementsModule.forChild(),
{ provide: LazyElementsStateService, useClass: CustomStateService }
]
```

right ?

Reply:
Already now you could `{ provide: LazyElementsLoaderService, useClass: CustomLazyElementsLoaderService }` but yeah that would a bit more implementation compared to just extracting state management into stand-alone service